interfaces/acados_template/acados_template/mpc_utils.py

```diff
@@ -72,10 +390,52 @@ def create_model_with_cost_state(ocp: AcadosOcp) -> Tuple[AcadosModel, np.ndarra
     elif ocp.cost.cost_type == "CONVEX_OVER_NONLINEAR":
         cost_dot = ca.substitute(
             model.cost_psi_expr, model.cost_r_in_psi_expr, model.cost_y_expr)
-
     else:
         raise Exception("create_model_with_cost_state: Unknown cost type.")
 
+    return cost_dot
+
+def get_terminal_cost_expression(ocp: AcadosOcp):
+    model = ocp.model
+    if ocp.cost.cost_type_e == "LINEAR_LS":
+        y = ocp.cost.Vx_e @ model.x
+        residual = y - ocp.cost.yref
+        cost_dot = 0.5*(residual.T @ ocp.cost.W_e @ residual)
+
+    elif ocp.cost.cost_type == "NONLINEAR_LS":
+        residual = model.cost_y_expr_e - ocp.cost.yref_e
+        cost_dot = 0.5*(residual.T @ ocp.cost.W_e @ residual)
+
+    elif ocp.cost.cost_type == "EXTERNAL":
+        cost_dot = model.cost_expr_ext_cost_e
+
+    elif ocp.cost.cost_type == "CONVEX_OVER_NONLINEAR":
+        cost_dot = ca.substitute(
+            model.cost_psi_expr_e, model.cost_r_in_psi_expr_e, model.cost_y_expr_e)
```

Comment:
I think you need to substract `ocp.cost.yref_e` from the last argument, similarly for path cost